src/msg/async/Stack.cc

```diff
@@ -0,0 +1,191 @@
+      ldout(cct, 10) << __func__ << " starting" << dendl;
+      w->initialize();
+      w->init_done();
+      while (!w->done) {
+        ldout(cct, 30) << __func__ << " calling event process" << dendl;
+
+        int r = w->center.process_events(EventMaxWaitUs);
+        if (r < 0) {
+          ldout(cct, 20) << __func__ << " process events failed: "
+                         << cpp_strerror(errno) << dendl;
+          // TODO do something?
+        }
+      }
+      w->reset();
+    }
+  );
+}
+
+std::shared_ptr<NetworkStack> NetworkStack::create(CephContext *c, const string &t)
+{
+  if (t == "posix")
+    return std::shared_ptr<NetworkStack>(new PosixNetworkStack(c, t));
+
+  return nullptr;
+}
+
+Worker* NetworkStack::create_worker(CephContext *c, const string &type, unsigned i)
+{
+  if (type == "posix")
+    return new PosixWorker(c, i);
+  return nullptr;
+}
+
+NetworkStack::NetworkStack(CephContext *c, const string &t): type(t), started(false), cct(c)
+{
+  for (unsigned i = 0; i < cct->_conf->ms_async_max_op_threads; ++i) {
+    Worker *w = create_worker(cct, type, i);
+    workers.push_back(w);
+  }
+  num_workers = cct->_conf->ms_async_op_threads;
+}
+
+void NetworkStack::start()
+{
+  if (started)
+    return ;
+  simple_spin_lock(&pool_spin);
+  for (unsigned i = 0; i < num_workers; ++i)
+    add_thread(i);
+  spawn_workers(threads);
+  started = true;
+  simple_spin_unlock(&pool_spin);
+
+  for (unsigned i = 0; i < num_workers; ++i)
+    workers[i]->wait_for_init();
+}
+
+Worker* NetworkStack::get_worker()
+{
+  ldout(cct, 10) << __func__ << dendl;
+
+   // start with some reasonably large number
+  unsigned min_load = std::numeric_limits<int>::max();
+  Worker* current_best = nullptr;
+
+  simple_spin_lock(&pool_spin);
+  // find worker with least references
+  // tempting case is returning on references == 0, but in reality
+  // this will happen so rarely that there's no need for special case.
+  for (unsigned i = 0; i < num_workers; ++i) {
+    unsigned worker_load = workers[i]->references.load();
+    ldout(cct, 20) << __func__ << " Worker " << workers[i] << " load: " << worker_load << dendl;
+    if (worker_load < min_load) {
+      current_best = workers[i];
+      min_load = worker_load;
+    }
+  }
+
+  // if minimum load exceeds amount of workers, make a new worker
+  // logic behind this is that we're not going to create new worker
+  // just because others have *some* load, we'll defer worker creation
+  // until others have *plenty* of load. This will cause new worker
+  // to get assigned to all new connections *unless* one or more
+  // of workers get their load reduced - in that case, this worker
+  // will be assigned to new connection.
+  // TODO: add more logic and heuristics, so connections known to be
+  // of light workload (heartbeat service, etc.) won't overshadow
+  // heavy workload (clients, etc).
+  if (!current_best || ((threads.size() < (unsigned)cct->_conf->ms_async_max_op_threads)
```

Comment:
if there is at least one worker in `workers`. current_best won't be `nullptr`. so i am not sure if the policy described by the comment above would work as expected.